### ISSUE TITLE
vmac startup in standby state interrupted by IPv6

### ISSUE THREAD

**Issue by sim-** [acassen/keepalived#175]
Hello! I was trying vmac mode with v1.2.19, and ran into an issue. With vmac_xmit_base, state BACKUP, and IPv6 available, it emits a few ICMP6 messages from the virtual MAC on startup, which immediately causes the switches to learn the path to it. I don't see how this could work unless IPv6 is disabled or state MASTER on both sides. Without vmac_xmit_base, the elected master sends another VRRP frame fairly quickly and takes back the path, but this still causes some loss whenever a backup node comes online.

```
04:54:16.697013 00:00:5e:00:01:c6 > 33:33:00:00:00:16, ethertype IPv6 (0x86dd), length 90: :: > ff02::16: HBH ICMP6, multicast listener report v2, 1 group record(s), length 28
04:54:16.717322 00:1c:73:7d:f1:ed > 00:00:5e:00:01:c6, ethertype IPv4 (0x0800), length 98: 10.130.198.2 > 10.130.198.12: ICMP echo request, id 27189, seq 66, length 64
04:54:16.773015 00:00:5e:00:01:c6 > 33:33:00:00:00:16, ethertype IPv6 (0x86dd), length 90: :: > ff02::16: HBH ICMP6, multicast listener report v2, 1 group record(s), length 28
04:54:17.017058 00:00:5e:00:01:c6 > 33:33:ff:00:01:c6, ethertype IPv6 (0x86dd), length 78: :: > ff02::1:ff00:1c6: ICMP6, neighbor solicitation, who has fe80::200:5eff:fe00:1c6, length 24
04:54:18.017191 00:00:5e:00:01:c6 > 33:33:00:00:00:02, ethertype IPv6 (0x86dd), length 70: fe80::200:5eff:fe00:1c6 > ff02::2: ICMP6, router solicitation, length 16
04:54:18.717316 00:1c:73:7d:f1:ed > 00:00:5e:00:01:c6, ethertype IPv4 (0x0800), length 98: 10.130.198.2 > 10.130.198.12: ICMP echo request, id 27189, seq 68, length 64
04:54:19.717386 00:1c:73:7d:f1:ed > 00:00:5e:00:01:c6, ethertype IPv4 (0x0800), length 98: 10.130.198.2 > 10.130.198.12: ICMP echo request, id 27189, seq 69, length 64
04:54:20.717305 00:1c:73:7d:f1:ed > 00:00:5e:00:01:c6, ethertype IPv4 (0x0800), length 98: 10.130.198.2 > 10.130.198.12: ICMP echo request, id 27189, seq 70, length 64
```

**Comment by terryburton** [comment]
On hosts that do not require IPv6 I've worked around this by setting net.ipv6.conf.default.disable_ipv6 = 1 on startup. If you require IPv6 then you can firewall such messages when in BACKUP state using iptables, taking care to re-permit anything that is necessary as part of the MASTER notify script.

These solutions are far from ideal. Whilst in practise it currently happens to be the initial IPv6 control plane messages from the macvlan interface startup that leak the vmac on the BACKUP, it can in principle be any other control plane traffic (be it IPv6/v4 or something else) generated in response to certain system events (such as following live host migrations in VMware/Xen, etc.)

Bringing up the macvlan interface only when in MASTER state would seem to resolve the fundamental issue. Is anyone aware of some reason why this would be a bad idea? If not then can we consider this a feature request?

**Comment by pqarmitage** [comment]
There were several issues here. 

1. The link local address being created was based for the VMAC was based on the MAC address, and the VRRPv3 RFC specifically prohibits this. This meant that the link local address was the same on all keepalived routers, and hence the problems you observed with the switches. Commit 41426ec changed this so that the VMAC inherits the link local address from the underlying interface, if it has once, otherwise it creates one based on the MAC address of the underlying interface.

2. The MAC address used for IPv6 VRRP used to be wrong in keepalived; the penultimate octet should have been 0x02, rather than 0x01 (resolved in commit f4cab11 ). It may be better not to use vmac_xmit_base.

3. When creating a VMAC, unless IFLA_INET6_ADDR_GEN_MODE is set, a link local address is automatically generated for the interface based on the MAC address when the interface is brought up. Unfortunately this flag was only introduced in Linux 3.17, so for running on kernels earlier than this, as soon as the interface is created, the unwanted link local address is deleted (see commit 41426ec ).

The above commits should resolve your issues, principally by not duplicating the default link local address on multiple systems.
